File: server/services/order/internal/domain/order.go
package domain

import (
	"time"
)

// Order represents a booking order
type Order struct {
	ID             string `json:"id"`
	OrganizationID string `json:"organization_id"`
	UserID         string `json:"user_id"`
	TripID         string `json:"trip_id"`
	FromStationID  string `json:"from_station_id"`
	ToStationID    string `json:"to_station_id"`

	// Passengers
	Passengers []OrderPassenger `json:"passengers"`

	// Pricing
	SubtotalPaisa   int64  `json:"subtotal_paisa"`
	TaxPaisa        int64  `json:"tax_paisa"`
	BookingFeePaisa int64  `json:"booking_fee_paisa"`
	DiscountPaisa   int64  `json:"discount_paisa"`
	TotalPaisa      int64  `json:"total_paisa"`
	Currency        string `json:"currency"`

	// Payment
	PaymentID     string `json:"payment_id"`
	PaymentStatus string `json:"payment_status"`
	PaymentMethod string `json:"payment_method"`

	// Booking
	BookingID string       `json:"booking_id"`
	HoldID    string       `json:"hold_id"`
	Seats     []BookedSeat `json:"seats"`

	// Status
	Status OrderStatus `json:"status"`
	SagaID string      `json:"saga_id"`

	// Contact
	ContactEmail string `json:"contact_email"`
	ContactPhone string `json:"contact_phone"`

	// Timestamps
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	ExpiresAt time.Time `json:"expires_at"`

	// Idempotency
	IdempotencyKey string `json:"idempotency_key"`
}

type OrderStatus string

const (
	OrderStatusPending       OrderStatus = "pending"
	OrderStatusConfirmed     OrderStatus = "confirmed"
	OrderStatusFailed        OrderStatus = "failed"
	OrderStatusCancelled     OrderStatus = "cancelled"
	OrderStatusExpired       OrderStatus = "expired"
	OrderStatusRefundPending OrderStatus = "refund_pending"
	OrderStatusRefunded      OrderStatus = "refunded"
)

// OrderPassenger is a passenger on an order (distinct from anti-scalp Passenger)
type OrderPassenger struct {
	NID         string `json:"nid"`
	Name        string `json:"name"`
	SeatID      string `json:"seat_id"`
	SeatNumber  string `json:"seat_number"`
	SeatClass   string `json:"seat_class"`
	Gender      string `json:"gender"`
	Age         int    `json:"age"`
	NIDVerified bool   `json:"nid_verified"`
}

type BookedSeat struct {
	SeatID     string `json:"seat_id"`
	SeatNumber string `json:"seat_number"`
	SeatClass  string `json:"seat_class"`
	TicketID   string `json:"ticket_id"`
	PricePaisa int64  `json:"price_paisa"`
}

// PaymentStatus constants
const (
	PaymentStatusPending    = "pending"
	PaymentStatusAuthorized = "authorized"
	PaymentStatusCaptured   = "captured"
	PaymentStatusFailed     = "failed"
	PaymentStatusRefunded   = "refunded"
)

// CalculateTotals calculates order totals
func (o *Order) CalculateTotals(basePrices map[string]int64, taxRate float64, bookingFee int64) {
	o.SubtotalPaisa = 0
	for _, p := range o.Passengers {
		if price, ok := basePrices[p.SeatID]; ok {
			o.SubtotalPaisa += price
		}
	}

	o.TaxPaisa = int64(float64(o.SubtotalPaisa) * taxRate)
	o.BookingFeePaisa = bookingFee * int64(len(o.Passengers))
	o.TotalPaisa = o.SubtotalPaisa + o.TaxPaisa + o.BookingFeePaisa - o.DiscountPaisa

	if o.TotalPaisa < 0 {
		o.TotalPaisa = 0
	}
}

File: client/src/lib/api/order.ts
import { api } from './index';

export interface Passenger {
    nid: string;
    name: string;
    seat_id: string; // The seat ID they are occupying
    date_of_birth: string; // YYYY-MM-DD
    gender: string;
    age: number;
}

export interface PaymentMethod {
    type: string;
    token?: string;
}

export interface CreateOrderRequest {
    trip_id: string;
    from_station_id: string;
    to_station_id: string;
    hold_id: string;
    passengers: Passenger[];
    payment_method: PaymentMethod;
    contact_email: string;
    contact_phone: string;
    idempotency_key: string;
}

export interface Order {
    id: string;
    trip_id: string;
    route_id: string;
    from_station_id: string;
    to_station_id: string;
    status: string; // "PENDING", "CONFIRMED", "CANCELLED"
    passengers: any[]; // mapped passenger structure
    subtotal_paisa: number;
    tax_paisa: number;
    booking_fee_paisa: number;
    discount_paisa: number;
    total_paisa: number;
    currency: string;
    payment_status: string;
    contact_email: string;
    contact_phone: string;
    created_at: string;
    expires_at: string;
}

export const orderApi = {
    createOrder: async (payload: CreateOrderRequest): Promise<Order> => {
        const response = await api.post<Order>('/v1/orders', payload);
        return response;
    },

    getOrder: async (orderId: string): Promise<Order> => {
        const response = await api.get<Order>(`/v1/orders/${orderId}`);
        return response;
    },

    listOrders: async (pageToken?: string): Promise<{ orders: Order[], next_page: string, total: number }> => {
        const query = pageToken ? `?page_token=${pageToken}` : '';
        const response = await api.get<{ orders: Order[], next_page: string, total: number }>(`/v1/orders${query}`);
        return response;
    },

    cancelOrder: async (orderId: string, reason: string): Promise<{ success: boolean, order: Order, refund?: any }> => {
        const response = await api.post<{ success: boolean, order: Order, refund?: any }>(`/v1/orders/${orderId}/cancel`, { reason });
        return response;
    }
};
